### PR TITLE
Add Base64Url support for netstandard 2.0

### LINKED ISSUE #1658: Support Base 64 URL
Updated by @MihaZupan on 2024-02-27

## Proposed API

```c#
namespace System.Buffers.Text;

public static class Base64Url
{
    // Encode bytes => utf8
    public static OperationStatus EncodeToUtf8(ReadOnlySpan<byte> bytes, Span<byte> utf8, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true);
    public static OperationStatus EncodeToUtf8InPlace(Span<byte> buffer, int dataLength, out int bytesWritten);

    // Decode from utf8 => bytes
    public static OperationStatus DecodeFromUtf8(ReadOnlySpan<byte> utf8, Span<byte> bytes, out int bytesConsumed, out int bytesWritten, bool isFinalBlock = true);
    public static OperationStatus DecodeFromUtf8InPlace(Span<byte> buffer, out int bytesWritten);

    // Max length APIs
    public static int GetMaxDecodedFromUtf8Length(int length);
    public static int GetMaxEncodedToUtf8Length(int length);

    // IsValid
    public static bool IsValid(ReadOnlySpan<char> base64UrlText);
    public static bool IsValid(ReadOnlySpan<char> base64UrlText, out int decodedLength);
    public static bool IsValid(ReadOnlySpan<byte> base64UrlTextUtf8);
    public static bool IsValid(ReadOnlySpan<byte> base64UrlTextUtf8, out int decodedLength);

    // Up to this point, this is a mirror of System.Buffers.Text.Base64
    // Below are more helpers that bring over functionality similar to Convert.*Base64*

    // Encode to / decode from chars
    public static bool TryEncodeToChars(ReadOnlySpan<byte> bytes, Span<char> chars, out int charsWritten) { }
    public static bool TryDecodeFromChars(ReadOnlySpan<char> chars, Span<byte> bytes, out int bytesWritten) { }


    // These are just accelerator methods.
    // Should be efficiently implementable on top of the other ones in just a few lines.

    // Encode to string
    public static string EncodeToString(ReadOnlySpan<char> chars, Encoding encoding) { }
    public static string EncodeToString(ReadOnlySpan<byte> bytes) { }

    // Decode from chars => string
    // Decode from chars => byte[]
    // The names could also just be "Decode" without naming the return type
    public static string DecodeToString(ReadOnlySpan<char> chars, Encoding encoding) { }
    public static byte[] DecodeToByteArray(ReadOnlySpan<char> chars) { }
}
```

---

## Original issue

The [Base64](../blob/master/src/libraries/System.Memory/src/System/Buffers/Text/Base64.cs) implementation in **System.Memory** provides excellent low-level optimizations for [RFC 4648](https://tools.ietf.org/html/rfc4648), but it currently uses a fixed alphabet. Minor refactoring would add additional use cases using the existing implementation.

## Rationale and Usage
The most obvious use case for this change is to support the encoding variant known as _Base 64 URL_ also described in [RFC 4648 §4](https://tools.ietf.org/html/rfc4648#section-4).

This excerpt from the RFC describes the difference between the standard Base 64 alphabet and the Base 64 URL alphabet.

>_This encoding is technically identical to the previous one, except
   for the 62:nd and 63:rd alphabet character..._

I have already been able to verify the encoding and decoding in a copy of the current implementation by merely changing the 2 relevant characters in the alphabet mappings at:

- [Line 519](../blob/master/src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs#L519) in **Base64Encoder.cs**
- [Line 611](../blob/master/src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs#L611) in **Base64Decoder.cs**

Today, this logic is suboptimally implemented in at least:

- [Base64UrlTextEncoder.cs](../blob/master/src/Http/WebUtilities/src/Base64UrlTextEncoder.cs) (ASP.NET Core)
- [WebEncoders.cs](../blob/master/src/Shared/WebEncoders/WebEncoders.cs) (ASP.NET Core)

Furthermore, the encoding is generic and has use cases outside of ASP.NET (e.g. you shouldn't have to reference ASP.NET to use it).

## Proposed API Change
The existing `Base64.EncodeToUtf8` and `Base64.DecodeFromUtf8` should each add a new method overload that allows one of the following:

1. A new enumeration of allowed alphabets (ex: `Base64Alphabet`) which internally maps to well-known alphabet spans

2. Allow any custom alphabet to be supplied as `ReadOnlySpan<sbyte>` that **must** have an exact length of 64 for encoding and 256 for decoding

    a. One or more new types _could_ be provided with static properties for the well-known alphabet spans

## Details

Option 1 requires less validation, but has less flexibility. Option 2 has more flexibility, including scenarios not described here, but requires more validation before using the alphabet.

Although the standard Base 64 and Base 64 URL are technically the same encoding with different alphabets, Base 64 URL typically does not include padding. [RFC 4648 §3.2](https://tools.ietf.org/html/rfc4648#section-3.2) indicates this is allowed (as it's explicitly stated), but there is no need to make that concession in this API. Including the `=` character for padding in Base 64 URL encoding is still correct.

Both approaches would have a similar looking API:

```c#
public static unsafe OperationStatus EncodeToUtf8(
    ReadOnlySpan<byte> bytes,
    Span<byte> utf8,
    Base64Alphabet alphabet,
    out int bytesConsumed,
    out int bytesWritten,
    bool isFinalBlock = true);

public static unsafe OperationStatus DecodeFromUtf8(
    ReadOnlySpan<byte> utf8,
    Span<byte> bytes,
    Base64Alphabet alphabet,
    out int bytesConsumed,
    out int bytesWritten,
    bool isFinalBlock = true);
```
_Figure 1: Supply an alphabet enumeration_

```c#
public static unsafe OperationStatus EncodeToUtf8(
    ReadOnlySpan<byte> bytes,
    Span<byte> utf8,
    ReadOnlySpan<byte> alphabet,
    out int bytesConsumed,
    out int bytesWritten,
    bool isFinalBlock = true);

public static unsafe OperationStatus DecodeFromUtf8(
    ReadOnlySpan<byte> utf8,
    Span<byte> bytes,
    ReadOnlySpan<byte> alphabet,
    out int bytesConsumed,
    out int bytesWritten,
    bool isFinalBlock = true);
```
_Figure 2: Supply a custom alphabet_

The support for trimming off and re-adding padding should be implemented separately. This _could_ be in a separate `Base64Url` class or, perhaps more appropriately, added as a new type in **System.Text.Encodings.Web**.

Padding is generally very cheap to deal with compared to other implementation methods. Trimming involves walking the tail end of the span while there are padding characters and then slicing it off. Re-padding fills the end of the span buffer before decoding. Neither operation requires additional allocations.

## Open Questions

- Does Option 1 or Option 2 make more sense?
- To complete the cycle, it _feels_ like there should be a type that fully handles the encoding and decoding with trimmed padding. Does that make sense to be in **System.Memory**, **System.Text.Encodings.Web**, or perhaps some other place?

### PR DESCRIPTION
**Pull request by buyaa-n**
Add Base64Url package that supports netstandard 2.0 and .NET framework

Pretty much same implementation excluding the parts that not supported in netstandard like:
- No vectorization, though the serial part of the implementation kept as is
- No static abstract interface, therefore, no code sharing between Span<char>, Span<byte> overloads
- No `Span.IndexOfAnyExcept(SearchValues)` support so validation APIs logic differs around that are, but main logic is same  
- No Math.DivRem available

Fixes https://github.com/dotnet/runtime/issues/1658

### CODE REVIEW

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
Comment (ViktorHofer) [inline src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:6]: Why disable the compiler generated doc xml file?

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
Comment (ViktorHofer) [inline src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:22]: I _think_ we should just include the "..ref/Microsoft.Bcl.Memory.Forwards.cs" file here instead of setting this property. @ericstj opinions?

File: src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj
Comment (ViktorHofer) [inline src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:1]: I understand that you follow the pattern from existing Bcl projects but I wonder why we need this reference source project and compile it, given that we don't ship it in the package. That seems like unnecessary work to me.

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
Comment (ericstj) [reply src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:6]: Yep - keep it enabled.  Best to use the source-authored docs here.

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
Comment (ericstj) [reply src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:22]: Correct, it's better to hand author the forwards here.  No need to use partial facade generation.

File: src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj
Comment (ericstj) [reply src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:1]: Seems reasonable to support a case where we don't have the ref project - but we'd want to make sure nothing else in the build system depends on it.

**Comment by ViktorHofer** [comment]
@buyaa-n I tried my feedback out in a code space. Here's the patch that you can directly apply:

```patch
From 2b35645c7eee21dfdbdd35855d709c9def195097 Mon Sep 17 00:00:00 2001
From: Viktor Hofer <viktor.hofer@microsoft.com>
Date: Tue, 18 Jun 2024 16:17:54 +0000
Subject: [PATCH] Enable compiler doc generation and simplify forwards

---
 .../src/Microsoft.Bcl.Memory.csproj                  | 12 +++++++-----
 1 file changed, 7 insertions(+), 5 deletions(-)

diff --git a/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj b/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
index aac0fba3d95..fa7bc6e2c33 100644
--- a/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
+++ b/src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
@@ -3,7 +3,6 @@
   <PropertyGroup>
     <TargetFrameworks>netstandard2.0;$(NetFrameworkMinimum);$(NetCoreAppCurrent)</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
-    <UseCompilerGeneratedDocXmlFile>false</UseCompilerGeneratedDocXmlFile>
     <IsPackable>true</IsPackable>
     <!-- Disabling baseline validation since this is a brand new package.
          Once this package has shipped a stable version, the following line
@@ -19,17 +18,20 @@
 
   <!-- DesignTimeBuild requires all the TargetFramework Derived Properties to not be present in the first property group. -->
   <PropertyGroup>
-    <IsPartialFacadeAssembly Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</IsPartialFacadeAssembly>
-    <OmitResources Condition="'$(IsPartialFacadeAssembly)' == 'true'">true</OmitResources>
+    <OmitResources Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">true</OmitResources>
   </PropertyGroup>
 
-  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
+    <Compile Include="..\ref\Microsoft.Bcl.Memory.Forwards.cs" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
     <Compile Include="System\Buffers\Text\Base64UrlDecoder.cs" />
     <Compile Include="System\Buffers\Text\Base64UrlEncoder.cs" />
     <Compile Include="System\Buffers\Text\Base64UrlValidator.cs" />
   </ItemGroup>
 
-  <ItemGroup Condition="'$(IsPartialFacadeAssembly)' != 'true'">
+  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
     <PackageReference Include="System.Memory" Version="$(SystemMemoryVersion)" />
   </ItemGroup>

```

File: src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj
Comment (ViktorHofer) [reply src/libraries/Microsoft.Bcl.Memory/ref/Microsoft.Bcl.Memory.csproj:1]: Let's follow-up on that in a separate issue.

**Comment by ViktorHofer** [comment]
The allconfigurations leg is failing because the netstandard2.0 version of Base64Url has the ClsCompliant(false) attribute on it but the net9.0 version doesn't. If that's intentional, create a suppression file. The error explains how to do that.

File: src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj
Comment (buyaa-n) [reply src/libraries/Microsoft.Bcl.Memory/src/Microsoft.Bcl.Memory.csproj:6]: Sure, just copied from existing Microsoft.Bcl.* packages

**Comment by ViktorHofer** [comment]
No, I don't have write permissions to your fork. You can either enable perms or apply the patch yourself (if `git apply ...` doesn't work, feel free to manually update the file).

**Comment by buyaa-n** [comment]
> No, I don't have write permissions to your fork. You can either enable perms

For the PR `Allow edits and access to secrets by maintainers` checkbox is checked, I will try to enable perms on my fork 

> apply the patch yourself (if `git apply ...` doesn't work, feel free to manually update the file).

I tried that and `error: corrupt patch at line 45`

**Comment by buyaa-n** [comment]
> No, I don't have write permissions to your fork. You can either enable perms

@ViktorHofer I added you as collaborator, please accept and apply your patch

**Comment by ViktorHofer** [comment]
The removal of the ClsCompliant(false) attribute is non-breaking, right? I'm 99% sure it isn't but just want to double check.

**Comment by buyaa-n** [comment]
> The removal of the ClsCompliant(false) attribute is non-breaking, right? I'm 99% sure it isn't but just want to double check.

Weird there were ClsCompliant warning on `uint` `ushort` types that is why I added the attribute, but somehow now I don't see the warning when removing the attribute, now removed the attribute and suppression

**Review by ViktorHofer — APPROVED** [review]
Infrastructure / project changes LGTM

File: src/libraries/Microsoft.Bcl.Memory/src/Resources/Strings.resx
Comment (stephentoub) [inline src/libraries/Microsoft.Bcl.Memory/src/Resources/Strings.resx:124]:
```suggestion
    <value>The input is not a valid Base64Url string as it contains a non-Base64Url character, more than two padding characters, or an illegal character among the padding characters.</value>
```

File: src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs
Comment (stephentoub) [inline src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:287]: There's a lot of code being duplicated here. Is there no opportunity for further code sharing with the .NET 9 implementation?

File: src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs
Comment (stephentoub) [inline src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:373]:
```suggestion
        private static bool IsValidPadding(uint padChar) => padChar is EncodingPadEqual or EncodingPadPercentage;
```
?

File: src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs
Comment (stephentoub) [inline src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:409]: This method will only ever be used with byte / char inputs, right? i.e. we don't need to care about int values > ushort.MaxValue? If so, I believe you could use this implementation, which is the 32-bit version of the 64-bit version used, and it would be appropriate for both 32-bit and 64-bit processes:
```C#
static bool IsWhiteSpace(int value)
{
    Debug.Assert(value >= 0 && value <= ushort.MaxValue);
    uint charMinusLowUInt32;
    return (int)((0xC8000100U << (short)(charMinusLowUInt32 = (ushort)(value - '\t'))) & (charMinusLowUInt32 - 32)) < 0;
}
```

**Review by stephentoub — APPROVED** [review]

File: src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs
Comment (buyaa-n) [reply src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:287]: Because cannot use static abstract methods in netstandard the main implementation cannot be shared, though I think some helper methods could be shared, assuming it will not need to add a netstandard target for CoreLib

Comment (stephentoub) [reply src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:287]: Would we be able to share it if we just used non-static abstract methods? e.g. instead have instance interface methods, implement those as instance methods on a struct, and pass in a default instance of the struct as an argument?

File: src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs
Comment (buyaa-n) [reply src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:409]: Yes, it is only used for byte char inputs, thanks!

File: src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs
Comment (buyaa-n) [reply src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:287]: I guess it could, but how that would affect performance?

Comment (stephentoub) [reply src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:287]: I'd hope very little, but you'd need to actually measure to see.

Comment (buyaa-n) [reply src/libraries/Microsoft.Bcl.Memory/src/System/Buffers/Text/Base64UrlDecoder.cs:287]: OK, I'll give it try

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlDecoder.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Url/Base64UrlDecoder.cs:24]: Having statics for these is a deoptimization. The call sites can just use eg `default(Base64UrlDecoderByte)`.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:115]: With the exception of the code in these ifdefs, and with calls being prefixed with `decoder.`, this whole file is an exact copy of what was in the other file, right? Or are there other changes?

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:652]: nit: indenting

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:1250]: Why is this 0?

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:1247]: Can we use `#if NET` instead of `#if NETCOREAPP` for all of these?

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64EncoderHelper.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64EncoderHelper.cs:17]: As commented earlier, all of these statically cached instances can be removed, and just use `default(T)` instead of reading this field.

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64EncoderHelper.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64EncoderHelper.cs:141]: Nit: indenting

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs
Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:115]: Exactly as you said,, only changes are if-defs for netsandard, and type parameter to `decoder` no new changes. Well, there is small bugfix that related to the new tests added, nothing else new

Basically all internal members from `Base64Decoder` and `Base64Encoder` moved in this `Base64Herlper` type to avoid conflict referencing `Base64` from netstandard project between .NET 9 and netstandard versions.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:115]: > Well, there is small bugfix that related to the new tests added, nothing else new

What is that fix?

Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:115]: I was fixed the bug found with Base64Url fuzzing only for byte overloads, the same bug left for char overloads.
 
The diff is here https://github.com/dotnet/runtime/pull/103617/commits/73acb51861f24f32ec1f2b3d2007486ca3ebbd42

File: src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs
Comment (buyaa-n) [reply src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Helper/Base64DecoderHelper.cs:1250]: Well, this source will not called for netsandard,  though I think now after splitting the Base64 public vs internal sources I don't need to do this.

**Pull request merged**
